src/execution_plan/optimizations/apply_join.c

```diff
@@ -166,14 +170,28 @@ void applyJoin(ExecutionPlan *plan) {
 			} else {
 				// The Cartesian Product still has a child operation; introduce the join op as another child.
 				ExecutionPlan_AddOp(cp, value_hash_join);
+				// Search for additional filters which can be solved by the join operation branches.
```

Comment:
As a general note,
This function is too long, please break it down such that each cartesian product see line 91 (main loop) is handled by a dedicated new function.